### PR TITLE
17191 fix is relative function

### PR DESCRIPTION
**Pull request by ggh2e3**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #17191 

I fixed the `BaseUrl::isRelative` method. I wrote some tests to cover all the  cases that came up to my mind.

Thanks

### CODE REVIEW

**Review by bizley — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by SamMousa** [comment]
@bizley just noted this today. This is the note from `parse_url`: 
> **Caution** This function may not give correct results for relative or invalid URLs, and the results may not even match common behavior of HTTP clients. If URLs from untrusted input need to be parsed, extra validation is required, e.g. by using [filter_var()](https://www.php.net/manual/en/function.filter-var.php) with the FILTER_VALIDATE_URL filter.

https://www.php.net/parse_url

This means that we're using `parse_url` incorrectly since we're expecting to use it on relative URLS.

**Comment by bizley** [comment]
Ugh, right. Looks like reverse is in order then. On it.

**Comment by bizley** [comment]
Hm, after some more research I think we can keep it - we rely here only on the scheme part (whether it's present or not) - the warning is for checking other url parts, there are known bugs with it, but I could not find something about scheme itself being not recognized properly.

**Comment by rob006** [comment]
BTW: benchmark that I mentioned in https://github.com/yiisoft/yii2/pull/20089#issuecomment-1871927855: https://3v4l.org/fNYfS 

`parse_url()` is 10x slower slower than `strpos()`, but it is still pretty fast.

**Comment by bizley** [comment]
I checked https://3v4l.org/ioiFVo - maybe we should go for triple strpos?

**Comment by rob006** [comment]
I would go with regexp. It is comparable to triple strpos in terms of performance for relative URLs (and faster for absolute URLs) and supports all protocols at the same time: https://3v4l.org/Vb0rJ

**Comment by SamMousa** [comment]
> Hm, after some more research I think we can keep it - we rely here only on the scheme part (whether it's present or not) - the warning is for checking other url parts, there are known bugs with it, but I could not find something about scheme itself being not recognized properly.

That's good news, so this is no longer a discussion about functional regression.

If it's just about performance it can lower the urgency a bit....

> I checked https://3v4l.org/ioiFVo - maybe we should go for triple strpos?

This is not a fair test; you're testing cases that are very favorable; either the pos is 0, or in case of `//` it's 4. The 3rd case is bad because it doesn't have contain it.
I've added `str_starts_with` which is even faster but not available on lower versions.

Btw: what do we do with mixed case URLs?


I'm working on a more detailed performance test; give me a few more mins!

**Comment by bizley** [comment]
Not sure what you mean by the fair test, I was testing whether the url starts with http:// or https:// or // hence the 0 (in real code this would require inversion of course). But I would go with the regex Rob proposed (`@^((\w+):)?//@`) since it's fast (and on top of everything it will take care of mixed case protocols :)

**Comment by SamMousa** [comment]
With not fair meant you're only testing cases where the URL is not relative.

https://3v4l.org/Dpob1

This is a cleaned up version if your original tests @bizley. 
- It has more examples, buth relative and non relative ones.
- It has 6 implementations

Output php 8.3.1:
 ```
preg_match: 0.168
preg_match (case insensitive): 0.173
strpos: 0.226
str_starts_with: 0.243
substr_compare (case insensitive): 0.204
substr_compare (case sensitive): 0.221
parse_url: 0.376
```

Output 7.4
```
preg_match: 0.181
preg_match (case insensitive): 0.182
strpos: 0.190
str_starts_with: 0.189
substr_compare (case insensitive): 0.133
substr_compare (case sensitive): 0.144
parse_url: 0.231
```

This is from doing 200.000 iterations for each of checking 6 examples. (Don't compare the times with your tests since I've added more boiler plate and am verifying the results).




Interestingly, as we're always told: don't optimize too early. As we see here at some point preg_match becomes faster, likely because the regex engine actually compiles the regular expression. This is very expensive but amortized over large number of iterations it becomes cheaper.

With 300.000 reps on 8.3.1 `preg_match` wins vs all other strategies.
```
preg_match: 0.184
preg_match (case insensitive): 0.172
strpos: 0.244
str_starts_with: 0.257
substr_compare (case insensitive): 0.219
substr_compare (case sensitive): 0.223
parse_url: 0.322
```
With 30.000 reps however (a lot less consistent) we see cases where it's twice as slow.
```
preg_match: 0.034
preg_match (case insensitive): 0.017
strpos: 0.025
str_starts_with: 0.027
substr_compare (case insensitive): 0.021
substr_compare (case sensitive): 0.023
parse_url: 0.033
```

Given these results I agree that we should just do preg match. It is the most readable code and works more than fast enough.


**TLDR: I agree with using `preg_match`**

**Comment by bizley** [comment]
Great work, guys! Let's go with preg_match then. @SamMousa you are already working on the PR, right?

**Comment by SamMousa** [comment]
Correct! ;-)  Just looking at the RFC for the scheme part:
```
 scheme        = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
```

https://www.rfc-editor.org/rfc/rfc3986

How'd you guess?
